hack/shoot-comparator/pkg/shoot/matcher.go

```diff
@@ -219,3 +168,189 @@ func (m *Matcher) NegatedFailureMessage(_ interface{}) string {
 func (m *Matcher) FailureMessage(_ interface{}) string {
 	return strings.Join(m.fails, "\n")
 }
+
+func addLabelsMatcher(e, a map[string]string, m *[]propertyMatcher) {
+	for key, val := range a {
+		*m = append(*m, propertyMatcher{
+			path:          fmt.Sprintf("metadata/labels/%s", key),
+			expected:      e,
+			GomegaMatcher: gomega.HaveKeyWithValue(key, val),
+		})
+	}
+}
+
+func addAnnotationsMatcher(e, a map[string]string, m *[]propertyMatcher) {
+	for key, val := range a {
+		*m = append(*m, propertyMatcher{
+			path:          fmt.Sprintf("metadata/annotations/%s", key),
+			expected:      e,
+			GomegaMatcher: gomega.HaveKeyWithValue(key, val),
+		})
+	}
+}
+
+func addDNSMatcher(e, a v1beta1.ShootSpec, m *[]propertyMatcher) {
+	*m = append(*m, propertyMatcher{
+		path:          "spec/dns",
+		expected:      e.DNS,
+		GomegaMatcher: newDNSMatcher(a),
+	})
+}
+
+func val(v interface{}) string {
+	if reflect.ValueOf(v).Kind() == reflect.Pointer && reflect.ValueOf(v).IsNil() {
+		return ""
+	}
+
+	if reflect.ValueOf(v).Kind() == reflect.Pointer {
+		return fmt.Sprintf("%v", reflect.ValueOf(v).Elem())
+	}
+
+	return fmt.Sprintf("%v", v)
+}
+
+func idToleration(v interface{}) string {
+	toleration, ok := v.(v1beta1.Toleration)
+	if !ok {
+		panic("invalid type")
+	}
+	return fmt.Sprintf("%s:%s", toleration.Key, val(toleration.Value))
+}
+
+func tolerations(ts []v1beta1.Toleration) gstruct.Elements {
+	out := map[string]types.GomegaMatcher{}
+	for _, t := range ts {
+		ID := idToleration(t)
+		out[ID] = gstruct.MatchAllFields(gstruct.Fields{
+			"Key":   gomega.BeComparableTo(t.Key),
+			"Value": gomega.BeComparableTo(t.Value),
+		})
+	}
+	return out
+}
+
+func idProvider(v interface{}) string {
+	provider, ok := v.(v1beta1.DNSProvider)
+	if !ok {
+		panic("invalid type")
+	}
+
+	return fmt.Sprintf("%s:%s:%s",
+		val(provider.Type),
+		val(provider.SecretName),
+		val(provider.Primary))
+}
+
+func providers(ps []v1beta1.DNSProvider) gstruct.Elements {
+	out := map[string]types.GomegaMatcher{}
+	for _, p := range ps {
+		ID := idProvider(p)
+
+		domainsMatcher := gomega.BeNil()
+		if p.Domains != nil {
+			domainsMatcher = gstruct.PointTo(gstruct.MatchFields(gstruct.IgnoreExtras, gstruct.Fields{
+				"Include": gomega.ContainElements(p.Domains.Include),
+			}))
+		}
+
+		out[ID] = gstruct.MatchFields(gstruct.IgnoreExtras, gstruct.Fields{
+			"Primary":    gomega.Equal(p.Primary),
```

Comment:
There is a bunch of fields that are set by KEB:
- `workers.maxSurge`
- `workers.maxUnavailable`
- `workers.maximum`
- `workers.minimum`
- `workers.name`
- `workers.volume.size`
- `workers.volume.type`
- `workers.zones`
- `workers.machine.type`

Those fields are [rewritten](https://github.com/kyma-project/infrastructure-manager/blob/e67746efebc95c821207a3e0c53d3c1e2fca1e8f/internal/gardener/shoot/extender/provider.go#L21) from Runtime CR into shoot CR. 

Open question: should we compare whole worker objects? Runtime CR uses `Worker` type from gardener so in theory someone could pass any e.g. `worker.taints` field.

Reply:
same issue as [here](https://github.com/kyma-project/infrastructure-manager/pull/418/files/bf3647e85b0f43426310f0bfc6e118a08c4d1142#r1801701656)

Reply:
> Open question: should we compare whole worker objects? Runtime CR uses Worker type from gardener so in theory someone could pass any e.g. worker.taints field.

ATM we only check for:
- providerConfig
- machine.image.name
- machine.image.version